File: app/src/main/java/com/pij/zworkout/workout/WorkoutDetailFragment.java
package com.pij.zworkout.workout;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.util.ObjectsCompat;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.annimon.stream.Optional;
import com.pij.zworkout.R;
import com.pij.zworkout.list.WorkoutsActivity;

import javax.inject.Inject;

import activitystarter.ActivityStarter;
import activitystarter.Arg;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnTextChanged;
import butterknife.Unbinder;
import dagger.android.support.DaggerFragment;
import io.reactivex.disposables.CompositeDisposable;

import static com.annimon.stream.Optional.ofNullable;

/**
 * A fragment representing a single Workout detail screen.
 * This fragment is either contained in a {@link WorkoutsActivity}
 * in two-pane mode (on tablets) or a {@link WorkoutDetailActivity}
 * on handsets.
 */
public class WorkoutDetailFragment extends DaggerFragment {
    private final CompositeDisposable subscriptions = new CompositeDisposable();
    /**
     * The fragment argument representing the item ID that this fragment
     * represents.
     */
    @Arg(optional = true)
    String itemId;
    @BindView(R.id.workout_detail)
    View workoutDetail;
    @BindView(R.id.name)
    TextView name;
    @BindView(R.id.description)
    TextView description;
    @Inject
    WorkoutViewModel viewModel;
    private Unbinder unbinder;

    public WorkoutDetailFragment() {
    }

    @NonNull
    public static WorkoutDetailFragment newInstance(Optional<String> itemId) {
        return itemId.map(WorkoutDetailFragmentStarter::newInstance)
                .orElse(WorkoutDetailFragmentStarter.newInstance());
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.workout_detail, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        ActivityStarter.fill(this, savedInstanceState);
        unbinder = ButterKnife.bind(this, view);
        setHasOptionsMenu(true);

        subscriptions.addAll(
                viewModel.model().map(Model::showError).filter(Optional::isPresent).map(Optional::get).subscribe(this::showError),
                viewModel.model().map(Model::inProgress).distinctUntilChanged().subscribe(this::showInProgress),
                viewModel.model().map(Model::name).distinctUntilChanged().subscribe(this::displayName),
                viewModel.model().map(Model::nameIsReadOnly).distinctUntilChanged().subscribe(this::disableName),
                viewModel.model().map(Model::description).distinctUntilChanged().subscribe(this::displayDescription)
        );

        if (savedInstanceState == null) {
            if (itemId == null) {
                viewModel.createWorkout();
            } else {
                viewModel.load(itemId);
            }
        }
    }

    @Override
    public void onDestroyView() {
        subscriptions.clear();
        unbinder.unbind();
        unbinder = null;
        super.onDestroyView();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.workout_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle item selection
        switch (item.getItemId()) {
            case R.id.menu_save:
                viewModel.save();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @OnTextChanged(R.id.name)
    void setName(CharSequence newValue) {
        viewModel.name(newValue.toString());
    }

    @OnTextChanged(R.id.description)
    void setDescription(CharSequence newValue) {
        viewModel.description(newValue.toString());
    }

    private void displayDescription(String newValue) {
        if (!ObjectsCompat.equals(description.getText().toString(), newValue)) {
            description.setText(newValue);
        }
    }

    private void disableName(boolean readOnly) {
        name.setEnabled(!readOnly);
    }

    private void displayName(String newValue) {
        if (!ObjectsCompat.equals(name.getText().toString(), newValue)) {
            name.setText(newValue);
        }
        // TODO put that in a different place in the lifecycle
        ofNullable(getActivity())
                .map(activity -> activity.findViewById(R.id.toolbar_layout))
                .map(toolbar -> (CollapsingToolbarLayout) toolbar)
                .ifPresent(appBarLayout -> appBarLayout.setTitle(newValue));
    }

    private void showInProgress(boolean inProgress) {
//        empty.setText(inProgress ? R.string.list_loading : R.string.list_empty);
    }

    private void showError(String message) {
        Snackbar.make(workoutDetail, message, Snackbar.LENGTH_LONG).show();
    }

}

File: app/src/main/java/com/pij/zworkout/workout/WorkoutDetailFragment.kt
/*
 * Copyright (c) 2018, Chiswick Forest
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and limitations under the License.
 */

package com.pij.zworkout.workout

import activitystarter.ActivityStarter
import activitystarter.Arg
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.view.*
import androidx.core.view.isVisible
import androidx.core.widget.toast
import com.pij.TextWatcherAdapter
import com.pij.updateEnabled
import com.pij.updateText
import com.pij.zworkout.R
import com.pij.zworkout.list.WorkoutsActivity
import dagger.android.support.DaggerFragment
import io.reactivex.disposables.CompositeDisposable
import kotlinx.android.synthetic.main.activity_workout_detail.*
import kotlinx.android.synthetic.main.fragment_workout.*
import javax.inject.Inject

/**
 * A fragment representing a single Workout detail screen.
 * This fragment is either contained in a [WorkoutsActivity]
 * in two-pane mode (on tablets) or a [WorkoutDetailActivity]
 * on handsets.
 */
class WorkoutDetailFragment : DaggerFragment() {
    private val subscriptions = CompositeDisposable()
    /**
     * The fragment argument representing the item ID that this fragment
     * represents.
     */
    @Arg(optional = true)
    var itemId: String? = null

    @Inject
    internal lateinit var viewModel: WorkoutViewModel

    @Inject
    internal lateinit var adapter: EffortsAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_workout, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        ActivityStarter.fill(this, savedInstanceState)
        setHasOptionsMenu(true)

        name.addTextChangedListener(TextWatcherAdapter { viewModel.name(it) })
        description.addTextChangedListener(TextWatcherAdapter { viewModel.description(it) })
        list.adapter = adapter

        subscriptions.add(viewModel.model().subscribe({ display(it) }, { it.printStackTrace(); activity?.toast("bad call!")?.show() }))

        if (savedInstanceState == null) {
            if (itemId == null) {
                viewModel.createWorkout()
            } else {
                viewModel.load(itemId!!)
            }
        }
    }

    private fun display(model: Model) {
        if (model.showError != null) {
            Snackbar.make(workout_detail, model.showError, Snackbar.LENGTH_LONG).show()
        }

        empty.setText(if (model.inProgress) R.string.detail_loading else R.string.detail_empty)
        name.updateText(model.name)
        toolbar_layout?.title = model.name
        name.updateEnabled(!model.nameIsReadOnly)
        description.updateText(model.description)
        list.isVisible = model.efforts.isNotEmpty()
        empty.isVisible = model.efforts.isEmpty()
        adapter.setItems(model.efforts)
    }

    override fun onDestroyView() {
        subscriptions.clear()
        super.onDestroyView()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.workout_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle item selection
        return when (item.itemId) {
            R.id.menu_save -> {
                viewModel.save()
                true
            }
            R.id.menu_add -> {
                viewModel.addEffort()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    companion object {

        fun newInstance(itemId: String?): WorkoutDetailFragment {
            return itemId?.let { WorkoutDetailFragmentStarter.newInstance(it) } ?: WorkoutDetailFragmentStarter.newInstance()
        }
    }

}

